src/utils.js

```diff
@@ -1,241 +1,227 @@
-		if (trigger.indexOf(special) > -1) {
-			return false;
-		}
-	}
-	return true;
-};
-
-/**
-string stringFormat (string type, string)
-
-Formats a string according to one of the following types:
-- formal
-- sentence
-- uppercase
-- lowercase
-*/
-exports.stringFormat = function(type, string) {
-	var first, i, len, result, word, words;
-	if (type === "uppercase") {
-		return string.toUpperCase();
-	} else if (type === "lowercase") {
-		return string.toLowerCase();
-	} else if (type === "sentence") {
-		string += "";
-		first = string.charAt(0).toUpperCase();
-		return first + string.substring(1);
-	} else if (type === "formal") {
-		words = string.split(/\s+/);
-		result = [];
-		for (i = 0, len = words.length; i < len; i++) {
-			word = words[i];
-			first = word.charAt(0).toUpperCase();
-			result.push(first + word.substring(1));
-		}
-		return result.join(" ");
-	}
-	return content;
-};
-
-/**
-[]string parseCallArgs
-
-Parse a string and return shell-like arguments as an array. Normally this
-means each word in the string becomes an item in the result, but quoted
-sections of the input will come back as a single item.
-
-Example:
-
-```javascript
-console.log( parseCallArgs('please google "writing chat bot"'));
-// ["please", "google", "writing chat bot"]
-```
-*/
-exports.parseCallArgs = function(str) {
-	var buff, c, doubleQuoteRe, flushBuffer, i, insideAString, len, result, spaceRe;
-	result = [];
-	buff = "";
-	insideAString = false;
-	spaceRe = /\s/ig;
-	doubleQuoteRe = /"/ig;
-	flushBuffer = function() {
-		if (buff.length !== 0) {
-			result.push(buff);
-		}
-		return buff = "";
-	};
-	for (i = 0, len = str.length; i < len; i++) {
-		c = str[i];
-		if (c.match(spaceRe) && !insideAString) {
-			flushBuffer();
-			continue;
-		}
-		if (c.match(doubleQuoteRe)) {
-			if (insideAString) {
-				flushBuffer();
-			}
-			insideAString = !insideAString;
-			continue;
-		}
-		buff += c;
-	}
-	flushBuffer();
-	return result;
-};
-
-/**
-object clone (object)
-
-Clone an object.
-*/
-exports.clone = function(obj) {
-	var copy, key;
-	if (obj === null || typeof obj !== "object") {
-		return obj;
-	}
-	copy = obj.constructor();
-	for (key in obj) {
-		if (!obj.hasOwnProperty(key)) {
-			continue;
-		}
-		copy[key] = exports.clone(obj[key]);
-	}
-	return copy;
-};
-
-/**
-int nIndexOf (string, string match, int index)
-
-Finds a match in a string at a given index
-
-Usage:
-string = "My name is Rive"
-match = " "
-index = 2
-return = 7
-
-Summary: It will look for a second space in the string
-*/
-exports.nIndexOf = function(string, match, index) {
-	return string.split(match, index).join(match).length;
-};
+ * Object clone (object)
+ *
+ * Clone an object.
+ */
+
+const clone = (obj) => {
+    if (obj === null || typeof obj !== "object") {
+      return obj
+    }
+    const copy = obj.constructor()
+    for (const key in obj) {
+      if (!obj.hasOwnProperty(key)) {
+        copy[key] = clone(obj[key])
+      }
+    }
+    return copy
+  },
```

Comment:
I get that these functions (`clone`, `extend`, `isAPromise`, etc.) are being declared as `const`s in series, but IMO it's a little confusing considering it's serial declaration over _hundreds of lines_ just for the sake of saving a few characters... for readability's sake, it might be nice to have these declared normally!

Reply:
I guess its more readable when its separated by modules, also makes thing more unix like, and make combinations and functional programming less difficulty. That's why thing are done this way. And utils.js is a module script, read a bit about ES2019, things are different in the future versions of JS. They maybe will put decorators!

Reply:
if you're looking to employ more a functional style, these methods could really benefit from refactoring to use `map`/`reduce`/`filter` instead of loops, and those iterator functions are conveniently part of ES6! :) To be honest, though, a lot of this code duplicates common (and, more importantly, _well-tested)_ functionality that could be imported from a library like Lodash or Underscore. If your goal is future maintainability, either of those options might help.

Reply:
If we aim to make more function, Ramda is the right choice.